tools/perfscale-load-generator/watcher/watcher.go

```diff
@@ -0,0 +1,231 @@
+			w.handleDeleted(obj)
+		},
+	})
+
+	return w
+}
+
+func (w *ObjListWatcher) Run() {
+	go w.informer.Run(w.stopChannel)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, w.informer.HasSynced) {
+		panic(fmt.Errorf("watcher timed out waiting for caches to sync"))
+	}
+}
+
+func (w *ObjListWatcher) Stop() {
+	close(w.stopChannel)
+}
+
+func (w *ObjListWatcher) handleUpdate(obj interface{}) {
+	switch w.ResourceKind {
+	case objUtils.VMIResource:
+		vmi, ok := obj.(*kvv1.VirtualMachineInstance)
+		if !ok {
+			log.Log.V(2).Errorf("Could not convert VirtualMachineInstance obj: %v", w.ResourceKind)
+		}
+		if vmi.Status.Phase == kvv1.Running {
+			w.addObj(string(vmi.GetUID()))
+		}
+
+	case objUtils.VMResource:
+		vm, ok := obj.(*kvv1.VirtualMachine)
+		if !ok {
+			log.Log.V(2).Errorf("Could not convert VirtualMachine obj: %v", w.ResourceKind)
+		}
+		if vm.Status.Ready {
+			w.addObj(string(vm.GetUID()))
+		}
+
+	case objUtils.VMIReplicaSetResource:
+		replicaSet, ok := obj.(*kvv1.VirtualMachineInstanceReplicaSet)
+		if !ok {
+			log.Log.V(2).Errorf("Could not convert VirtualMachineInstanceReplicaSet obj: %v", w.ResourceKind)
+		}
+		w.addObj(string(replicaSet.GetUID()))
+
+	default:
+		log.Log.V(2).Errorf("Watcher does not support object type %s", w.ResourceKind)
+		return
+	}
+}
+
+func (w *ObjListWatcher) handleDeleted(obj interface{}) {
+	switch w.ResourceKind {
+	case objUtils.VMIResource:
+		vmi, ok := obj.(*kvv1.VirtualMachineInstance)
+		if !ok {
+			log.Log.V(2).Errorf("Could not convert VirtualMachineInstance obj: %v", w.ResourceKind)
+		}
+		w.deleteObj(string(vmi.GetUID()))
+
+	case objUtils.VMResource:
+		vm, ok := obj.(*kvv1.VirtualMachine)
+		if !ok {
+			log.Log.V(2).Errorf("Could not convert VirtualMachine obj: %v", w.ResourceKind)
+		}
+		w.deleteObj(string(vm.GetUID()))
+
+	case objUtils.VMIReplicaSetResource:
+		replicaSet, ok := obj.(*kvv1.VirtualMachineInstanceReplicaSet)
+		if !ok {
+			log.Log.V(2).Errorf("Could not convert VirtualMachineInstanceReplicaSet obj: %v", w.ResourceKind)
+		}
+		w.deleteObj(string(replicaSet.GetUID()))
+
+	default:
+		log.Log.V(2).Errorf("Watcher does not support object type %s", w.ResourceKind)
+		return
+	}
+}
+
+func (w *ObjListWatcher) WaitRunning(timeout time.Duration) error {
+	if w.getRunningCount() == w.desiredObjRunningCount {
+		return nil
+	}
+	for {
+		select {
+		case <-time.After(timeout):
+			return fmt.Errorf("timeout after %s waiting for objects be running", timeout)
+
+		case <-w.updateChannel:
+			if w.getRunningCount() == w.desiredObjRunningCount {
+				return nil
+			}
+		}
+	}
+}
+
+func (w *ObjListWatcher) WaitDeletion(timeout time.Duration) error {
+	if w.getRunningCount() == 0 {
+		return nil
+	}
+	for {
+		select {
+		case <-time.After(timeout):
+			return fmt.Errorf("timeout after %s waiting for objects be deleted", timeout)
```

Comment:
Will leftover VMIs ruin our test or future tests?

Reply:
Yes, because the VMI names might collide. The test configuration will have a namespace prefix, so the namespace and VMI names will always repeat. The error will be related to creating VMIs with the same name. 
However, the test will never count leftover VMIs from previous test execution because each execution has a Label with UUID and the tool uses this label to watch and delete objects.

If someone decides to run the tests without clean up (e.g., for debugging purposes), they must manually clean up the tests later.

Also note that, if the test exits unexpectedly, like with a crtl-c, the test will clean up and delete all created namespaces.

Reply:
Cleanup can be a corner cases with certain types of tests, but I don't think we're ready for those yet.  We can iterate on this over time.